# **MAE598 Homework 1**

### Q1. Solve the following problem using Python SciPy.optimize or JuMP. Please attach your code and results. Specify your initial guesses of the solution. If you change your initial guess, do you find different solutions?


###Minimize:
(x1-x2)^2+(x2+x3-2)^2+(x4-1)^2+(x5-1)^2
###Subject to:
x1+3x2 = 0
x3+x4-2x5 = 0
x2-x5 = 0
###Boundary Conditions :
-10<= xi <= 10, i = 1,2,3,4,5
###Initial guesses of the solution
# [-4,-5,-6,-7,-8],[8,7,6,5,4],[20,21,22,23,24],[30,30,30,30,30],[-30,-30,-30,-30,-30],[0,0,0,0,0]

In [16]:

# First Iteration with initial conditions [-4,-5,-6,-7,-8]
from scipy.optimize import minimize

# Given Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2 + (x[3]-1)**2 + (x[4]-1)**2

# Given Constraints
constraint = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Given Boundary Conditions
boundary = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
result1 = minimize(fun, (-4,-5,-6,-7,-8), method='SLSQP', bounds=boundary,constraints=constraint)

# Output
result1
print(result1)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 4.093023311079211
       x: [-7.675e-01  2.558e-01  6.281e-01 -1.164e-01  2.558e-01]
     nit: 7
     jac: [-2.047e+00 -1.855e-01 -2.232e+00 -2.233e+00 -1.488e+00]
    nfev: 42
    njev: 7


In [10]:
# Second Iteration with initial conditions [8,7,6,5,4]
from scipy.optimize import minimize

# Given Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2 + (x[3]-1)**2 + (x[4]-1)**2

# Given Constraints
constraint = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Given Boundary Conditions
boundary = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Given Execution
result2 = minimize(fun, (8,7,6,5,4), method='SLSQP', bounds=boundary,constraints=constraint)

# Output
result2
print(result2)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 4.09302325586575
       x: [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
     nit: 7
     jac: [-2.047e+00 -1.860e-01 -2.233e+00 -2.233e+00 -1.488e+00]
    nfev: 43
    njev: 7


In [11]:
# Third Iteration with initial conditions [20,21,22,23,24]
from scipy.optimize import minimize

# Given Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2 + (x[3]-1)**2 + (x[4]-1)**2

# Given Constraints
constraint = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Given Boundary Conditions
boundary = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Given Execution
result3 = minimize(fun, (20,21,22,23,24), method='SLSQP', bounds=boundary,constraints=constraint)

# Output
result3
print(result3)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 4.093023258012615
       x: [-7.675e-01  2.558e-01  6.279e-01 -1.162e-01  2.558e-01]
     nit: 8
     jac: [-2.047e+00 -1.861e-01 -2.233e+00 -2.232e+00 -1.488e+00]
    nfev: 48
    njev: 8


In [12]:
# Fourth Iteration with initial conditions [30,30,30,30,30]
from scipy.optimize import minimize

# Given Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2 + (x[3]-1)**2 + (x[4]-1)**2

# Given Constraints
constraint = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Given Boundary Conditions
boundary = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
result4 = minimize(fun, (30,30,30,30,30), method='SLSQP', bounds=boundary,constraints=constraint)

# Output
result4
print(result4)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 4.093023258012615
       x: [-7.675e-01  2.558e-01  6.279e-01 -1.162e-01  2.558e-01]
     nit: 8
     jac: [-2.047e+00 -1.861e-01 -2.233e+00 -2.232e+00 -1.488e+00]
    nfev: 48
    njev: 8


In [13]:
# Fifth Iteration with initial conditions [-30,-30,-30,-30,-30]
from scipy.optimize import minimize

# Given Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2 + (x[3]-1)**2 + (x[4]-1)**2

# Given  Constraints
constraint = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Given Boundary Conditions
boundary = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
result5 = minimize(fun, (-30,-30,-30,-30,-30), method='SLSQP', bounds=boundary,constraints=constraint)

# Output
result5
print(result5)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 4.09302325604787
       x: [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
     nit: 8
     jac: [-2.047e+00 -1.860e-01 -2.233e+00 -2.233e+00 -1.488e+00]
    nfev: 48
    njev: 8


In [14]:
# Sixth Iteration with initial conditions [0,0,0,0,0]
from scipy.optimize import minimize

# Given Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2 + (x[3]-1)**2 + (x[4]-1)**2

# Given Constraints
constraint = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Given Boundary Conditions
boundary = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Execution
result6 = minimize(fun, (0,0,0,0,0), method='SLSQP', bounds=boundary,constraints=constraint)

# Output
result6
print(result6)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 4.093023255813954
       x: [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
     nit: 3
     jac: [-2.047e+00 -1.860e-01 -2.233e+00 -2.233e+00 -1.488e+00]
    nfev: 20
    njev: 3


###If you change your initial guess, do you find different solutions?


Initial condition (Iteration) | Function | X values
--- | --- | ---
[-4,-5,-6,-7,-8] | 4.093023311079211 |  [-7.675e-01  2.558e-01  6.281e-01 -1.164e-01  2.558e-01]
[8,7,6,5,4] | 4.09302325586575 |  [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
[20,21,22,23,24] | 4.093023258012615 |  [-7.675e-01  2.558e-01  6.279e-01 -1.162e-01  2.558e-01]
[30,30,30,30,30] | 4.093023258012615 |  [-7.675e-01  2.558e-01  6.279e-01 -1.162e-01  2.558e-01]
[-30,-30,-30,-30,-30] | 4.09302325604787 |  [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
[0,0,0,0,0] | 4.093023255813954 |  [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]



From the above comparison, it can be seen that over a large range of inital conditions, there is very minimum change in the solution.


### Q2. Let x and b ∈ Rn be vectors and A ∈ Rn×n be a square matrix. Define
f : Rn → R as f (x) = b^T x + x^T Ax.

###Answer -

(a) What are the gradient and Hessian of f (x) with respect to x?

Gradient g(x)=∂f(x)/∂x = b+(A+A^T)x ; Hessian H(x)= ∂g(x)/∂x = (A+A^T)


(b) Derive the first- and second-order Taylor’s approximations of f (x)
at x = 0. Are these approximations exact?


The first-order Taylor's approximation of a function f(x) at a point x = a is given by:
f(x) ≈ f(a) + ∇f(a)^T (x - a)

The second-order Taylor's approximation includes both linear and quadratic terms:
f(x) ≈ f(a) + ∇f(a)^T (x - a) + 0.5 * (x - a)^T H(a) (x - a)


At x = 0:

First-order Taylor's approximation:
f(x) ≈ f(0) + ∇f(0)^T x = b^T x

Second-order Taylor's approximation:
f(x) ≈ f(0) + ∇f(0)^T x + 0.5 * x^T H(0) x = b^T x + 0.5 * x^T (A + A^T) x


Are these approximations exact?

No, these approximations are not exact. They are valid when x is close to the point of approximation (x = 0 in this case). The higher-order terms in the Taylor's series expansion are neglected in these approximations, making them accurate only in the vicinity of the point of approximation





### Q3. Let A ∈ Rn×n be a square matrix

### Answer-

a) What are the necessary and sufficient conditions for A to be pos-
itive definite?

If all of the eigen values are positive, then A is positive definite.

(b) What are the necessary and sufficient conditions for A to have full
rank?

If the determinant of A is nonzero then A will have full rank.

(c) If there exists y ∈ Rn and y 6 = 0 such that AT y = 0, then what
are the conditions on b for Ax = b to have a solution for x?

b should be orthogonal to y for Ax=b to have a solution for x.



### Q4. Due to the recent inflation, let’s reconsider the Stigler diet problem proposed by Nobel laureate George Stigler after the second World War: Consider that the grocery store offers N types of food of your interest, and each food contains the same M types of nutrition. Denote aij as the quantity of nutrition type j of food type i for i = 1, 2, ..., N and j = 1, 2, ..., M , ci the unit price of food type i, and bj the necessary quantity of nutrition type j for a month. Formulate an optimization problem to determine the minimum grocery cost to satisfy the nutrition needs.

### Answer-

Variables:

Let's define the decision variables:
xi as the quantity of food type i to be purchased.

Objective Fuction:

The objective is to minimize the total cost while meeting the nutritional requirements:

Minimize:
$\sum_{i=1}^N c_{i} x_{i}$   

Constraints:

1) Nutritional Requirements Constraint:
For each nutritional type j, the total amount of nutrition obtained from the selected foods should meet or exceed the required amount $b_{j}$

$\sum_{i=1}^N a_{ij} x_{i} >= b_{j}$, for $j=1,2,...,M$


2) Non-Negativity Constraint:
The quantity of each food type purchased cannot be negative:
$x_{i} >= 0$, for $i = 1,2,...,N$

The complete optimization problem can be summarized as:

Minimize: $\sum_{i=1}^N c_{i} x_{i}$  

Subject to:

$\sum_{i=1}^N a_{ij} x_{i} >= b_{j}$, for $j=1,2,...,M$

$x_{i} >= 0$, for $i = 1,2,...,N$


In this problem, $N$ represents the number of food types available, $M$ represents the number of nutritional types, $a_{ij}$  represents the quantity of nutritional type $j$ in food type $i$, $c_{i}$ represents the cost of food type $i$, and $ b_{j}$ represents the required quantity of nutritional type $j$ for a month.

